### PR TITLE
chore: add unit tests for Request classes

### PR DESCRIPTION
**Pull request by Mohammad-Dwairi**
# Situation

With the introduction of the new SDK core, we need to raise the tests coverage before attempting to migrate the core module to the original repository. We will be testing the new core with this LC SDK during this temporary phase.

# Task

1. Cover `Request` HTTP model class with necessary unit tests
2. Cover all related HTTP model classes with unit tests (e.g. `RequestBody`, `Protocol`, `Status`, `Method`, etc...) 

# Action

1. Cleaned up the `Request` class from unused functionalities and improved the structure.
2. Added missing `@JvmStatic` and `JvmOverloads` annotations where needed. 
3. Dropped the custom `Url` class and used native Java `URL` for simplicity.
4. Added Kotlin & Java tests for:
    - `Request.kt`
    - `RequestBody.kt`
    - `Protocol.kt`
    - `Method.kt`
    - `Status.kt`
5. Fixed some classes that were affected by changes made to these classes

# Testing

All tests are passing

# Results

All classes tested in the PR are 100% covered
Screenshot 2024-12-05 at 00 03 51

### CODE REVIEW

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/StatusTest.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+class StatusTest {
+
+    @Test
+    fun `should return correct Status for valid codes`() {
+        assertEquals(Status.CONTINUE, Status.fromCode(100))
+        assertEquals(Status.OK, Status.fromCode(200))
+        assertEquals(Status.MOVED_PERMANENTLY, Status.fromCode(301))
+        assertEquals(Status.BAD_REQUEST, Status.fromCode(400))
+        assertEquals(Status.INTERNAL_SERVER_ERROR, Status.fromCode(500))
+        assertEquals(Status.THIS_IS_FINE, Status.fromCode(218))
```
Comment (jordan-n-schmidt) [inline code/src/test/kotlin/com/expediagroup/sdk/core/http/StatusTest.kt:16]: Do you want to add the rest?

File: code/src/test/java/com/expediagroup/sdk/core/http/RequestBodyJavaTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        InputStream inputStream = new ByteArrayInputStream(content.getBytes());
+        RequestBody.create(inputStream);
+    }
+
+    @Test
+    @DisplayName("should access RequestBody.create as static method with InputStream, media type, content length")
+    public void shouldCreateRequestBodyFromInputStreamAndMediaTypeAndContentLength() {
+        String content = "Hello World";
+        InputStream inputStream = new ByteArrayInputStream(content.getBytes());
+        RequestBody.create(inputStream, MediaType.Companion.getTEXT_PLAIN(), content.getBytes().length);
```
Comment (mohnoor94) [inline code/src/test/java/com/expediagroup/sdk/core/http/RequestBodyJavaTest.java:27]: Can you please annotate the properties in the MediaType companion object with `@JvmField` so we can call them like `MediaType.TEXT_PLAIN` from Java?

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+        // When
+        val request = Request
+            .builder()
+            .method(method)
+            .url(url)
+            .addHeader(headerName, headerValue)
+            .build()
+
+        // When
```
Comment (mohnoor94) [inline code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt:95]: Expect*

Please also update it for the rest of the cases in the file.

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/StatusTest.kt (hunk 1 above)
Comment (Mohammad-Dwairi) [reply code/src/test/kotlin/com/expediagroup/sdk/core/http/StatusTest.kt:16]: I'll add common ones, but it's too much to cover all of them. they're unlikely to be used by any API..

File: code/src/test/java/com/expediagroup/sdk/core/http/RequestBodyJavaTest.java (hunk 2 above)
Comment (Mohammad-Dwairi) [reply code/src/test/java/com/expediagroup/sdk/core/http/RequestBodyJavaTest.java:27]: This was addressed in PR #117 
Will fix everything related to `MediaType` once we merge it

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt (hunk 3 above)
Comment (Mohammad-Dwairi) [reply code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt]: Done

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/StatusTest.kt (hunk 1 above)
Comment (Mohammad-Dwairi) [reply code/src/test/kotlin/com/expediagroup/sdk/core/http/StatusTest.kt]: Done!

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        val request = Request
+            .builder()
+            .method(method)
+            .url(url)
+            .setHeader(headerName, headerValue1)
+            .setHeader(headerName, headerValue2)
+            .build()
+
+        // When
```
Comment (mohnoor94) [inline code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt:159]: Sorry for the nitpicking, you missed updating this.

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            .builder()
+            .method(Method.GET)
+            .url("https://example.com")
+            .addHeader("Header1", "Value1")
+            .setHeader("Header2", "Value2")
+            .setHeader("Header1", "NewValue1")
+            .addHeader("Header3", listOf("Value1", "Value2"))
+            .build()
+
+        // When & Expect
```
Comment (mohnoor94) [inline code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt:219]: I think the `when` part is merged with `given` above, not with `expect`, Same for other test cases.

**Review by mohnoor94 — CHANGES_REQUESTED** [review]

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt (hunk 4 above)
Comment (Mohammad-Dwairi) [reply code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt]: Fixed!

File: code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt (hunk 5 above)
Comment (Mohammad-Dwairi) [reply code/src/test/kotlin/com/expediagroup/sdk/core/http/RequestTest.kt]: Done! https://github.com/ExpediaGroup/lodging-connectivity-java-sdk/pull/118/commits/34b3ad93ec49e1f5c370c83167561e0fb1592856

**Review by mohnoor94 — DISMISSED** [review]

**Review by mohnoor94 — APPROVED** [review]

**Review by mohnoor94 — APPROVED** [review]

**Review by jordan-n-schmidt — APPROVED** [review]
Status.THIS_IS_FINE

**Pull request merged**